### PR TITLE
HHH-16209 Identically-named association in entity root and embeddable leads to mixup during association loading

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-16209

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java [hunk 1]
```diff
@@ -205,6 +205,30 @@ public boolean isSuffix(DotIdentifierSequence dotIdentifierSequence) {
...
+	 * Removes the dotIdentifierSequence part from the NavigablePath,
+	 * when the NavigablePath does not contain the dotIdentifierSequence it returns null;
+	 *
+	 * @param dotIdentifierSequence the part to remove from the NavigablePath
+	 *
+	 * @return the NavigablePath stripped of the dotIdentifierSequence part
+	 * or null if the NavigablePath does not contain the dotIdentifierSequence.
+	 *
+	 */
+	public NavigablePath withoutDotIdentifierSequence(DotIdentifierSequence dotIdentifierSequence) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java:219]:
```suggestion
	public NavigablePath trimSuffix(DotIdentifierSequence dotIdentifierSequence) {
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java [hunk 2]
```diff
@@ -1065,7 +1065,15 @@ class Mother {
 			}
 			return false;
 		}
-		return parentNavigablePath.isSuffix( bidirectionalAttributePath );
+
+		NavigablePath navigablePath = parentNavigablePath.withoutDotIdentifierSequence( bidirectionalAttributePath );
+		if ( navigablePath != null ) {
+			if ( navigablePath.getLocalName().equals( EntityIdentifierMapping.ROLE_LOCAL_NAME ) ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java:1071]: Could this also be a foreign key role?

**Review by beikov — COMMENTED** [review]
LGTM, just a small suggestion

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java (hunk 2 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java:1071]: I thought about it , probably not but not 100% sure

File: hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java]: thanks @beikov name changed

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java (hunk 2 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java:1071]: do you thing it is safer to add
```
final String localName = navigablePath.getLocalName();
			if ( localName.equals( EntityIdentifierMapping.ROLE_LOCAL_NAME )
					|| localName.equals( ForeignKeyDescriptor.PART_NAME )
					|| localName.equals( ForeignKeyDescriptor.TARGET_PART_NAME ) ) {
				navigablePath = navigablePath.getParent();
			}
```
?

**Pull request merged**